README.md
<img src='https://firebasestorage.googleapis.com/v0/b/misdevelop.appspot.com/o/my_device%2FPackages%20pub.dev%204.png?alt=media&token=7567b5b0-4c7c-48d7-8d7d-adcf794fe8c3' alt='My Device from MIS Develop'>
  
## Features
    
Agile & easy handler for device information 📱

Includes Device Info Plus ℹ️ and Connectivity Plus 🛜 implementations

## Getting started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  my_device: ^0.2.0
```

In your library add the following import:

```dart
import 'package:my_device/my_device.dart';
```

Once context is initialized, do the following to initialize the DeviceManager instance:

```dart
DeviceManager(context);
```

You can also use the configure method to initialize the DeviceManager instance, responsive breakpoints and Connectivity configuration:

```dart
DeviceManager.configure(
context: context,
responsiveBreakpoints: const ResponsiveBreakpoints(
  mobile: 400, //Default is 500
  desktop: 1000, //Default is 900
),
connectivityConfigure: ConnectivityConfigure(
  onConnectivityChanged: (connectivity) {
    print('Connectivity changed: $connectivity');
  },
  onConnectionLost: () {
    print('Connection lost');
  },
  onConnectionReestablished: () {
    print('Connection reestablished');
  },
  useDefaultBehaviour: true // Shows a default toast when set to true
),
);
// Mobile view < mobile
// Tablet view > mobile && < desktop
// Desktop view > desktop
```

## Usage
### Any method can be accessed through the DeviceManager instance or through the `context` extension
#### Getters for screen height and width
```dart
final screenHeight = DeviceManager.screenHeight;
final screenWidth = DeviceManager.screenWidth;
final height = DeviceManager.height;
final width = DeviceManager.width;
final heightFromContext = context.height;
final widthFromContext = context.width;
```

#### Getters for responsive sizes (dependent of the configured responsive breakpoints)
```dart
final isPhoneView = DeviceManager.isPhoneView; // width < mobile
final isTabletView = DeviceManager.isTabletView; // mobile <= width < desktop
final isDesktopView = DeviceManager.isDesktopView; // width >= desktop
```

#### Simple methods for platform information
```dart
final isWeb = DeviceManager.isWeb;
final isAndroid = DeviceManager.isAndroid;
final isIOS = DeviceManager.isIOS;
final isMacOS = DeviceManager.isMacOS;
final isLinux = DeviceManager.isLinux;
final isWindows = DeviceManager.isWindows;

final platformName = DeviceManager.platformName;
final platformIcon = DeviceManager.platformIcon;
```

#### Direct method for device default padding
```dart
final padding = DeviceManager.viewPadding;
final paddingFromContext = context.viewPadding;
```

#### Connectivity Plus methods
##### Get the current status
```dart
final connectivity = DeviceManager.connectivity;
final connectivityFromContext = context.connectivity;
```
##### Listen on connectivity changes
```dart
DeviceManager.listenConnectivityChanges();
```
##### Set any custom connectivity method after configure callback
```dart
DeviceManager.onConnectionLost = () {
  print('Connection lost');
};

DeviceManager.onConnectionReestablished = () {
  print('Connection reestablished');
};

DeviceManager.onConnectivityChanged = (connectivity) {
  print('Connectivity changed: $connectivity');
};
```

#### Device Info Plus shortened method
 
```dart
final deviceInfo = DeviceManager.deviceInfo;
```


   
## Additional information
This package assumes corresponding permissions depending on platform.
Also take a look at the example app for further information.

CHANGELOG.md
## 0.2.1

- Improved documentation

## 0.2.0

- Included extension to use any method through `context`
- Responsive breakpoints and Connectivity configuration
- Added functions and improved documentation
- Updated to the latest version of Device Info Plus and Connectivity Plus

## 0.1.0

- Updated to the latest version of Device Info Plus and Connectivity Plus

## 0.0.9

- Added support for the latest version of Connectivity

## 0.0.8
MacOS fixes

## 0.0.7
 Web fixes
 
 ## 0.0.6
 Web bug fixes
 
 ## 0.0.5
 Improved README and example
 
## 0.0.4
 - Device instance getter
 - Fix import condition
 
## 0.0.3
 - Device Info Plugin support
 - Direct access to instance

## 0.0.2
Minor fixes

## 0.0.1
First implementation of my_device

lib/src/device.dart
import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:device_info_plus/device_info_plus.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:my_device/src/connectivity_configure.dart';
import 'package:my_device/src/platform.dart';
import 'package:my_device/src/responsive_breakpoints.dart';

///Access to device information and internet connection
class DeviceManager {
  static bool get isDesktopView => width >= responsiveBreakpoints.desktop;
  static bool get isTabletView =>
      width >= responsiveBreakpoints.mobile &&
      width < responsiveBreakpoints.desktop;
  static bool get isPhoneView => width < responsiveBreakpoints.mobile;
  static double get width => MediaQuery.of(context).size.width;
  static double get height => MediaQuery.of(context).size.height;

  ///Responsive breakpoints
  ResponsiveBreakpoints _responsiveBreakpoints = const ResponsiveBreakpoints();

  ///Is native Web
  bool _isWeb = false;

  ///Is native Android
  bool _isAndroid = false;

  ///Is native IOS
  bool _isIOS = false;

  ///Is native MacOS
  bool _isMacOS = false;

  ///Is native Windows
  bool _isWindows = false;

  ///Is native Linux
  bool _isLinux = false;

  ///Must be initialized every time you need to refresh the information
  late BuildContext _context;

  ///All information from Device Info Plugin
  Map<String, dynamic>? _deviceInfo;

  ///Internet connection state
  ///Has 3 types of connection:
  ///Wifi, mobile, none
  ConnectivityResult? _connectivity;

  ///Connectivity configuration
  ConnectivityConfigure? _connectivityConfigure;

  // Getters

  ///Device screen height
  static double get screenHeight => MediaQuery.of(context).size.height;

  ///Device screen width
  static double get screenWidth => MediaQuery.of(context).size.width;

  ///Is Web environment
  static bool get isWeb => DeviceManager.i._isWeb;

  ///Is native Android
  static bool get isAndroid => DeviceManager.i._isAndroid;

  ///Is native IOS
  static bool get isIOS => DeviceManager.i._isIOS;

  ///Is native MacOS
  static bool get isMacOS => DeviceManager.i._isMacOS;

  ///Is native Windows
  static bool get isWindows => DeviceManager.i._isWindows;

  ///Is native Linux
  static bool get isLinux => DeviceManager.i._isLinux;

  ///Default padding of handle bars, notches and others
  static EdgeInsets get viewPadding => MediaQuery.of(context).viewPadding;

  ///Device MediaQuery easy access
  static MediaQueryData get mediaQuery => MediaQuery.of(context);

  ///All information from Device Info Plugin
  static Map<String, dynamic>? get deviceInfo => DeviceManager.i._deviceInfo;

  ///Internet connection state
  ///Has 3 types of connection:
  ///Wifi, mobile, none
  static ConnectivityResult? get connectivity => DeviceManager.i._connectivity;

  ///context getter
  static BuildContext get context => DeviceManager.i._context;

  ///Responsive breakpoints
  static ResponsiveBreakpoints get responsiveBreakpoints =>
      DeviceManager.i._responsiveBreakpoints;

  ///Current platform name
  static String get platformName {
    if (isWeb) return 'Web';
    if (isAndroid) return 'Android';
    if (isIOS) return 'IOS';
    if (isMacOS) return 'MacOS';
    if (isWindows) return 'Windows';
    if (isLinux) return 'Linux';
    return 'Unknown';
  }

  ///Current platform icon
  static IconData get platformIcon {
    if (isWeb) return Icons.web_rounded;
    if (isAndroid) return Icons.android_rounded;
    if (isIOS) return Icons.apple_rounded;
    if (isMacOS) return Icons.desktop_mac_rounded;
    if (isWindows) return Icons.laptop_windows_rounded;
    if (isLinux) return Icons.laptop_chromebook_rounded;
    return Icons.help_rounded;
  }

  ///Shorter access to instance
  static DeviceManager get device => DeviceManager.i;

  ///Shorter access to instance
  static DeviceManager get i => instance;

  //Setters

  ///On connection lost callback
  static set onConnectionLost(Function callback) {
    instance._connectivityConfigure =
        instance._connectivityConfigure?.copyWith(onConnectionLost: callback) ??
            ConnectivityConfigure(onConnectionLost: callback);
  }

  ///On connection reestablished callback
  static set onConnectionReestablished(Function callback) {
    instance._connectivityConfigure = instance._connectivityConfigure
            ?.copyWith(onConnectionReestablished: callback) ??
        ConnectivityConfigure(onConnectionReestablished: callback);
  }

  ///On connectivity changed callback
  static set onConnectivityChanged(
      Function(ConnectivityResult result) callback) {
    instance._connectivityConfigure = instance._connectivityConfigure
            ?.copyWith(onConnectivityChanged: callback) ??
        ConnectivityConfigure(onConnectivityChanged: callback);
  }

  ///Use default behavior
  static set useDefaultBehavior(bool value) {
    instance._connectivityConfigure =
        instance._connectivityConfigure?.copyWith(useDefaultBehavior: value) ??
            ConnectivityConfigure(useDefaultBehavior: value);
  }

  ///Hold all device information
  static final DeviceManager instance = DeviceManager._internal();

  DeviceManager._internal();

  ///Internet connection alerts
  static setConnectivity(ConnectivityResult connect) {
    if (connect == ConnectivityResult.none &&
        instance._connectivity != ConnectivityResult.none) {
      if (instance._connectivityConfigure?.useDefaultBehavior ?? false) {
        showText(context, "No internet connection", '',
            seconds: 4, backgroundColor: Colors.amber[800]);
      } else {
        instance._connectivityConfigure?.onConnectionLost?.call();
      }
    }

    if (connect != ConnectivityResult.none &&
        instance._connectivity == ConnectivityResult.none) {
      if (instance._connectivityConfigure?.useDefaultBehavior ?? false) {
        showText(context, "Connection reestablish!", '',
            seconds: 2, backgroundColor: Colors.green[800]);
      } else {
        instance._connectivityConfigure?.onConnectionReestablished?.call();
      }
    }

    instance._connectivity = connect;
  }

  static void configure({
    BuildContext? context,
    ResponsiveBreakpoints? responsiveBreakpoints,
    ConnectivityConfigure? connectivityConfigure,
  }) async {
    await _init(context, responsiveBreakpoints, connectivityConfigure);
    listenConnectivityChanges();
  }

  static void listenConnectivityChanges() {
    Connectivity().onConnectivityChanged.listen((ConnectivityResult result) {
      setConnectivity(result);
      instance._connectivityConfigure?.onConnectivityChanged?.call(result);
    });
  }

  static Future<void> _init(
    BuildContext? context,
    ResponsiveBreakpoints? responsiveBreakpoints,
    ConnectivityConfigure? connectivityConfigure,
  ) async {
    if (context != null) instance._context = context;
    if (responsiveBreakpoints != null)
      instance._responsiveBreakpoints = responsiveBreakpoints;
    if (connectivityConfigure != null)
      instance._connectivityConfigure = connectivityConfigure;
    instance._isWeb = kIsWeb;
    instance._isAndroid = kIsWeb ? false : Platform.isAndroid;
    instance._isIOS = kIsWeb ? false : Platform.isIOS;
    instance._isMacOS = kIsWeb ? false : Platform.isMacOS;
    instance._isWindows =
        kIsWeb ? false : TargetPlatform.windows == defaultTargetPlatform;
    instance._isLinux =
        kIsWeb ? false : TargetPlatform.linux == defaultTargetPlatform;
    getInfo();
    instance._connectivity = await Connectivity().checkConnectivity();
  }

  factory DeviceManager(
    context, {
    ResponsiveBreakpoints? responsiveBreakpoints,
    ConnectivityConfigure? connectivityConfigure,
  }) {
    _init(context, responsiveBreakpoints, connectivityConfigure);
    return instance;
  }
  static Map<String, dynamic> _readWebBrowserInfo(WebBrowserInfo data) {
    return <String, dynamic>{
      'browserName': data.browserName.name,
      'appCodeName': data.appCodeName,
      'appName': data.appName,
      'appVersion': data.appVersion,
      'deviceMemory': data.deviceMemory,
      'language': data.language,
      'languages': data.languages,
      'platform': data.platform,
      'product': data.product,
      'productSub': data.productSub,
      'userAgent': data.userAgent,
      'vendor': data.vendor,
      'vendorSub': data.vendorSub,
      'hardwareConcurrency': data.hardwareConcurrency,
      'maxTouchPoints': data.maxTouchPoints,
    };
  }

  static getInfo() async {
    instance._deviceInfo = await getDeviceInfo();
  }

  ///Device info depending on the actual platform
  static Future<Map<String, dynamic>> getDeviceInfo() async {
    Map<String, dynamic> deviceInfoData = {};
    DeviceInfoPlugin deviceInfo = DeviceInfoPlugin();

    deviceInfoData['lastConnection'] = DateTime.now().toString();

    if (kIsWeb) {
      WebBrowserInfo webBrowserInfo = await deviceInfo.webBrowserInfo;
      if (kDebugMode) {
        print('Running on ${webBrowserInfo.userAgent}');
      }
      deviceInfoData['infoDevice'] = _readWebBrowserInfo(webBrowserInfo);
      deviceInfoData['id'] = webBrowserInfo.userAgent;
    } else {
      if (Platform.isAndroid) {
        AndroidDeviceInfo androidInfo = await deviceInfo.androidInfo;
        if (kDebugMode) {
          print('Running on ${androidInfo.model}');
        }

        deviceInfoData['infoDevice'] = androidInfo.data;
        deviceInfoData['id'] = androidInfo.id;
      } else if (Platform.isIOS) {
        IosDeviceInfo iosInfo = await deviceInfo.iosInfo;

        if (kDebugMode) {
          print('Running on ${iosInfo.data}');
        }

        deviceInfoData['infoDevice'] = iosInfo.data;
        deviceInfoData['id'] = iosInfo.identifierForVendor;
      } else if (Platform.isMacOS) {
        MacOsDeviceInfo macInfo = await deviceInfo.macOsInfo;

        if (kDebugMode) {
          print('Running on ${macInfo.data}');
        }

        deviceInfoData['infoDevice'] = macInfo.data;
        deviceInfoData['id'] = macInfo.systemGUID;
      }
    }
    return deviceInfoData;
  }

  static void showText(BuildContext context, String title, String subtitle,
      {required int seconds, Color? backgroundColor}) {
    //Show dialog
    ScaffoldMessenger.of(context).showSnackBar(SnackBar(
      content: Column(
        children: [
          Text(title,
              style:
                  const TextStyle(fontWeight: FontWeight.bold, fontSize: 18)),
          Text(
            subtitle,
            style: const TextStyle(fontSize: 14),
          ),
        ],
      ),
      duration: Duration(seconds: seconds),
      backgroundColor: backgroundColor,
    ));
  }
}

extension DeviceExtension on BuildContext {
  ///Device screen height
  double get screenHeight => DeviceManager.screenHeight;

  ///Device screen width
  double get screenWidth => DeviceManager.screenWidth;

  ///Device screen height
  double get height => DeviceManager.height;

  ///Device screen width
  double get width => DeviceManager.width;

  ///Is Web environment
  bool get isWeb => DeviceManager.isWeb;

  ///Is native Android
  bool get isAndroid => DeviceManager.isAndroid;

  ///Is native IOS
  bool get isIOS => DeviceManager.isIOS;

  ///Is native MacOS
  bool get isMacOS => DeviceManager.isMacOS;

  ///Is native Windows
  bool get isWindows => DeviceManager.isWindows;

  ///Is native Linux
  bool get isLinux => DeviceManager.isLinux;

  ///Default padding of handle bars, notches and others
  EdgeInsets get viewPadding => DeviceManager.viewPadding;

  ///Device MediaQuery easy access
  MediaQueryData get mediaQuery => DeviceManager.mediaQuery;

  ///All information from Device Info Plugin
  Map<String, dynamic>? get deviceInfo => DeviceManager.deviceInfo;

  ///Internet connection state
  ///Has 3 types of connection:
  ///Wifi, mobile, none
  ConnectivityResult? get connectivity => DeviceManager.connectivity;

  ///Shorter access to instance
  DeviceManager get device => DeviceManager.i;
}


lib/src/connectivity_configure.dart
import 'package:connectivity_plus/connectivity_plus.dart';

class ConnectivityConfigure {
  /// Custom method to be called when the connection is changed
  final Function(ConnectivityResult result)? onConnectivityChanged;

  /// Custom method to be called when the connection is lost
  final Function? onConnectionLost;

  /// Custom method to be called when the connection is reestablished
  final Function? onConnectionReestablished;

  /// If true, a toast will be shown when the connection is lost and reestablished.
  final bool useDefaultBehavior;

  const ConnectivityConfigure({
    this.onConnectivityChanged,
    this.onConnectionLost,
    this.onConnectionReestablished,
    this.useDefaultBehavior = false,
  });

  /// Copy with method
  ConnectivityConfigure copyWith({
    Function(ConnectivityResult result)? onConnectivityChanged,
    Function? onConnectionLost,
    Function? onConnectionReestablished,
    bool? useDefaultBehavior,
  }) {
    return ConnectivityConfigure(
      onConnectivityChanged:
          onConnectivityChanged ?? this.onConnectivityChanged,
      onConnectionLost: onConnectionLost ?? this.onConnectionLost,
      onConnectionReestablished:
          onConnectionReestablished ?? this.onConnectionReestablished,
      useDefaultBehavior: useDefaultBehavior ?? this.useDefaultBehavior,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:my_device/my_device.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'My Device',
      theme: ThemeData(
        primarySwatch: Colors.deepPurple,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    Key? key,
  }) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    // Set the responsive breakpoints
    DeviceManager.configure(
      context: context,
      responsiveBreakpoints: const ResponsiveBreakpoints(
        mobile: 500,
        desktop: 1000,
      ),
      connectivityConfigure: ConnectivityConfigure(
        onConnectivityChanged: (connectivity) {
          print('Connectivity changed: $connectivity');
        },
        onConnectionLost: () {
          print('Connection lost');
        },
        onConnectionReestablished: () {
          print('Connection reestablished');
        },
      ),
    );

    // Or set a custom function after the configure
    DeviceManager.onConnectionLost = () {
      print('Connection lost');
    };

    DeviceManager.onConnectionReestablished = () {
      print('Connection reestablished');
    };

    DeviceManager.onConnectivityChanged = (connectivity) {
      print('Connectivity changed: $connectivity');
    };

    //Get the MediaQuery from the given context
    final mediaQuery = DeviceManager.mediaQuery;

    // Get the device screen height
    final screenHeight = DeviceManager.screenHeight;

    // Get the device screen width
    final screenWidth = DeviceManager.screenWidth;

    // You can also use the context directly to get width and height
    final width = context.width;
    final height = context.height;

    // Padding from physical hinges or islands
    var padding = DeviceManager.viewPadding;

    return Scaffold(
      appBar: AppBar(
        title: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('My device - ${DeviceManager.platformName}  '),
            Icon(DeviceManager.platformIcon, color: Colors.black),
          ],
        ),
      ),
      backgroundColor: const Color(0xFFc1c1c1),
      body: Stack(
        children: <Widget>[
          Padding(padding: padding),
          Flex(
            direction: Axis.vertical,
            children: [
              Expanded(
                flex: 0,
                child: Padding(
                  padding: const EdgeInsets.all(4.0),
                  child: Card(
                    child: Text(
                      'Device Pixel Ratio: ' + mediaQuery.devicePixelRatio.toString(),
                      textAlign: TextAlign.center,
                    ),
                  ),
                ),
              ), //Pixel ratio
              Wrap(
                alignment: WrapAlignment.center,
                runAlignment: WrapAlignment.center,
                spacing: 10,
                runSpacing: 10,
                children: [
                  SizedBox(
                    width: screenWidth / 2.3,
                    child: Card(
                        color: Colors.teal.withOpacity(.96),
                        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(20)),
                        child: Column(
                          mainAxisSize: MainAxisSize.min,
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            Text(
                              'Screen Width: ' + screenWidth.toString(),
                              textAlign: TextAlign.center,
                              style: const TextStyle(color: Colors.white),
                            ),
                            Text(
                              '\nScreen Height: ' + screenHeight.toString(),
                              textAlign: TextAlign.center,
                              style: const TextStyle(color: Colors.white),
                            ),
                            Text(
                              '\nDevice Orientation: ' + mediaQuery.orientation.toString(),
                              textAlign: TextAlign.center,
                              style: const TextStyle(color: Colors.white),
                            ),
                          ],
                        )),
                  ), //teal container
                  Column(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      SizedBox(
                        width: width / 2.3,
                        child: Card(
                          color: Colors.amber[700]!.withOpacity(.96),
                          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(20)),
                          child: Text(
                            'Connectivity: ' + DeviceManager.connectivity.toString(),
                            textAlign: TextAlign.center,
                            style: const TextStyle(color: Colors.white),
                          ),
                        ),
                      ),
                      SizedBox(
                        width: width / 2.3,
                        child: Card(
                          color: Colors.green[700]!.withOpacity(.96),
                          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(20)),
                          child: Column(
                            children: [
                              Text(
                                'View type: ' +
                                    (DeviceManager.isDesktopView
                                        ? 'Desktop'
                                        : DeviceManager.isTabletView
                                            ? 'Tablet'
                                            : 'Mobile'),
                                textAlign: TextAlign.center,
                                style: const TextStyle(color: Colors.white),
                              ),
                              const Text(
                                "\nBreakpoints>>>\nDesktop: 1000\nMobile: 500",
                                textAlign: TextAlign.center,
                                style: TextStyle(color: Colors.white),
                              ),
                            ],
                          ),
                        ),
                      ),
                    ],
                  ), //Yellow container
                ],
              ), //Teal & yellow containers
              SizedBox(
                height: height / 3,
                child: Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Card(
                      child: ListView(
                    children: [
                      displayEntry(context.deviceInfo!),
                    ],
                  )),
                ),
              ), //Device Info
            ],
          ),
        ],
      ),
    );
  }

  Widget displayEntry(Map e) {
    List<Widget> aux = [];
    for (var key in e.keys) {
      aux.add(buildEntry(key, e[key]));
    }
    return Column(children: aux);
  }
}

Widget buildEntry(key, e) {
  if (e is! Map) {
    if (e == null) {
      return ListTile(
        title: Text(key ?? ''),
      );
    }
    return ListTile(
      subtitle: Text(key ?? ''),
      title: Text(e.toString()),
    );
  } else {
    List<Widget> children = [];
    for (var key in e.keys) {
      children.add(buildEntry(key, e[key]));
    }
    return Column(children: children);
  }
}
